### ISSUE TITLE
Detect stale nonces

### ISSUE THREAD

**Issue by Sean-Der** [pion/turn#168]
Update `buildNonce` and `authenticateRequest` to generate and check the times on those.

**Comment by jech** [comment]
That's debatable.

RFC 8489 merely says "See Section 5.4 of [RFC7616] for guidelines.", while RFC 7616 says
```
   the server is
   free to construct the nonce such that it MAY only be used from a
   particular client, for a particular resource, for a limited period of
   time or number of uses, or any other restrictions.  Doing so
   strengthens the protection provided against, for example, replay
   attacks (see Section 5.5).  However, it should be noted that the
   method chosen for generating and checking the nonce also has
   performance and resource implications.
```

**Comment by Sean-Der** [comment]
@jech It also doesn't seem to be implemented correctly everywhere https://medium.com/confrere/gone-in-1100-seconds-hunting-bugs-on-the-edge-of-webrtc-132a186c45dd

I am afraid if I rolled it out in pion/turn it would lead to frustration/people having to turn it off. So for now I say we don't do this.

Unless someone more security minded comes by and tells me why it is a really bad idea :)

**Comment by fippo** [comment]
@Sean-Der that was the old Spartan Edge and its stack, don't make decisions based on that ;-)

**Comment by jech** [comment]
I'm really not too worried.  Even if stale nonces were to enable replay attacks (which is not at all obvious to me), they would only enable replay attacks against the TURN server, and the replayed traffic would still be rejected by the DTLS implementation at the other end.

**Comment by jech** [comment]
It looks like in fact we already implement nonce expiration:

  * https://github.com/pion/turn/blob/master/internal/server/nonce.go#L58
  * https://github.com/pion/turn/blob/master/internal/server/short_nonce.go#L116

It looks to me like the code in `short_nonce` is incorrect (it returns the wrong value, and doesn't use `nonceLifetime`).  @Sean-Der shall I file a new issue?

**Comment by Sean-Der** [comment]
Yes please file away @jech! I’m only paying attention to this repo will work on it right away
